ndb/src/google/cloud/ndb/eventloop.py

```diff
@@ -30,16 +31,228 @@
 ]
 
 
-def add_idle(*args, **kwargs):
-    raise NotImplementedError
+def _noop(*args, **kw):
+    """Do nothing."""
+
+# TODO: Use utils.logging_debug when implemented
+_logging_debug = _noop
+
+
+_Event = collections.namedtuple('_Event', (
+    'when', 'callback', 'args', 'kwargs'))
+
+
+class _Clock(object):
+    """A clock to determine the current time, in seconds."""
+
+    def now(self):
+        """Returns the number of seconds since epoch."""
+        return time.time()
+
+    def sleep(self, seconds):
+        """Sleeps for the desired number of seconds."""
+        time.sleep(seconds)
 
 
 class EventLoop:
-    __slots__ = ()
+    """Constructor.
+
+    Args:
+      clock: an eventloop._Clock object. Defaults to a time-based clock.
 
-    def __init__(self, *args, **kwargs):
+    Fields:
+      current: a FIFO list of (callback, args, kwds). These callbacks
+        run immediately when the eventloop runs.
+      idlers: a FIFO list of (callback, args, kwds). Thes callbacks
+        run only when no other RPCs need to be fired first.
+        For example, AutoBatcher uses idler to fire a batch RPC even before
+        the batch is full.
+      queue: a sorted list of (absolute time in sec, callback, args, kwds),
+        sorted by time. These callbacks run only after the said time.
+      rpcs: a map from rpc to (callback, args, kwds). Callback is called
+        when the rpc finishes.
+    """
+    __slots__ = ('clock', 'current', 'idlers', 'inactive', 'queue', 'rpcs')
+
+    def __init__(self, clock=None):
+        self.clock = clock if clock else _Clock()
+        self.current = collections.deque()
+        self.idlers = collections.deque()
+        self.inactive = 0
+        self.queue = []
+        self.rpcs = {}
+
+    def clear(self):
```

Comment:
This implementation is copied from the original source and seems kind of crazy to me.   I don't see how that while loop would ever run more than once, and the call to __init__ with the individual calls to clear the various structures which hang on to events seems redundant.   It looks like possibly someone was trying to code defensively for some kind of concurrent mutation situation on the task queues, but I'm not sure how that would actually come up given the event loop is supposed to run in a single thread.  I'd like to simplify this.